compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignTiles.cpp

```diff
@@ -122,289 +123,303 @@ LogicalResult duplicateGlobalObjFifos(RewriterBase &rewriter, Operation *op) {
-          return WalkResult::interrupt();
-        }
-        // Handle subviews.
+/// Base class for tile allocators to support different tile assignment
+/// strategies.
+class TileAllocatorBase {
+ public:
+  TileAllocatorBase(RewriterBase &rewriter,
+                    const AMDAIE::AMDAIEDeviceModel &deviceModel)
+      : rewriter(rewriter), deviceModel(deviceModel) {}
+  virtual ~TileAllocatorBase(){};
+  /// Assign tiles to the provided vector of logical objectFifos on the same
+  /// memory space. This method expects a set of valid tile candidates to be
+  /// provided inside the operation.
+  virtual LogicalResult assignTiles(
+      SmallVector<AMDAIE::LogicalObjFifoOpInterface> &objFifos,
+      uint8_t memSpace, function_ref<InFlightDiagnostic()> emitError) = 0;
+
+ protected:
+  RewriterBase &rewriter;
+  const AMDAIE::AMDAIEDeviceModel &deviceModel;
+};
+
+/// A custom tile allocater that takes into consideration the usage and column
+/// of users to determine tile locations.
+class UsageAndColumnBasedTileAllocator final : public TileAllocatorBase {
+ public:
+  DenseMap<Operation *, DenseSet<Operation *>> uniqueL3L2Pair;
+
+  UsageAndColumnBasedTileAllocator(
+      RewriterBase &rewriter, const AMDAIE::AMDAIEDeviceModel &deviceModel,
+      DenseMap<Operation *, DenseSet<Operation *>> uniqueL3L2Pair)
+      : TileAllocatorBase(rewriter, deviceModel),
+        uniqueL3L2Pair(uniqueL3L2Pair) {}
+
+  LogicalResult assignTiles(
+      SmallVector<AMDAIE::LogicalObjFifoOpInterface> &objFifos,
+      uint8_t memSpace, function_ref<InFlightDiagnostic()> emitError) {
+    assert(llvm::all_of(objFifos,
+                        [&](AMDAIE::LogicalObjFifoOpInterface objFifo) {
+                          return objFifo.getMemorySpaceAsUInt() == memSpace;
+                        }) &&
+           "All logical objectFifos should have ths same memory space");
+    if (memSpace == 2) return assignLocalTiles(objFifos, memSpace, emitError);
+    if (memSpace == 0 || memSpace == 1)
+      return assignNonLocalTiles(objFifos, memSpace, emitError);
+    return emitError() << "Unsupported memory space : "
+                       << std::to_string(memSpace);
+  }
+
+ private:
+  /// Assign tiles to the logical objectfifos with local memory space (L1).
+  /// The tiles are derived from the usage of the logical objectfifos within
+  /// core operations, which are already assigned a tile location.
+  LogicalResult assignLocalTiles(
+      SmallVector<AMDAIE::LogicalObjFifoOpInterface> &objFifos,
+      uint8_t memSpace, function_ref<InFlightDiagnostic()> emitError) {
+    assert(memSpace == 2 && "Local memory space should be `2`.");
+    for (AMDAIE::LogicalObjFifoOpInterface objFifo : objFifos) {
+      llvm::SmallSetVector<std::pair<int64_t, int64_t>, 16> tileLocations;
+      if (failed(findUsersInCoreAndAddTiles(
+              objFifo,
+              cast<AMDAIE::LogicalObjFifoOpInterface>(objFifo.getOperation()),
```

Comment:
```suggestion
              objFifo,
```

`objFifo` is already a `AMDAIE::LogicalObjFifoOpInterface`.